tools/models/model_factory.py

```diff
@@ -0,0 +1,168 @@
+            weight_kl=model_config.config["lambd_kl"],
+            device=rank,
+    )
+
+    # Initialize inner model
+    inner_model = INNModel(
+            ndim_tot=config["ndim_tot"],
+            ndim_x=config["ndim_x"],
+            ndim_y=config["ndim_y"],
+            ndim_z=config["ndim_z"],
+            loss_fit=fit,
+            loss_latent=MMD_multiscale,
+            loss_backward=MMD_multiscale,
+            lambd_predict=config["lambd_predict"],
+            lambd_latent=config["lambd_latent"],
+            lambd_rev=config["lambd_rev"],
+            zeros_noise_scale=config["zeros_noise_scale"],
+            y_noise_scale=config["y_noise_scale"],
+            hidden_size=config["hidden_size"],
+            activation=config["activation"],
+            num_coupling_layers=config["num_coupling_layers"],
+            device=rank,
+        )
+
+    # Initialize final model
+    model = ModelFinal(
+        VAE_obj,
+        inner_model,
+        EarthMoversLoss(),
+        weight_AE=config["lambd_AE"],
+        weight_IM=config["lambd_IM"],
+    )
+
+        # Load a pre-trained model
+    map_location = {"cuda:%d" % 0: "cuda:%d" % rank}
+    if config["load_model"] is not None:
+        original_state_dict = torch.load(
+            config["load_model"], map_location=map_location
+        )
+        # updated_state_dict = {key.replace('VAE.', 'base_network.'):
+        # value for key, value in original_state_dict.items()}
+        model.load_state_dict(original_state_dict)
+        print("Loaded pre-trained model successfully", flush=True)
+
+    elif config["load_model_checkpoint"] is not None:
+        model, _, _, _, _, _ = load_checkpoint(
+            config["load_model_checkpoint"],
+            model,
+            map_location=map_location,
+        )
+        print("Loaded model checkpoint successfully", flush=True)
+    else:
+        pass  # run with random init
+
+    lr = config["lr"]
+    bs_factor = (
+        io_config.trainBatchBuffer_config["training_bs"] / 2 * world_size
+    )
+    lr = lr * config["lr_scaling"](bs_factor)
+    print(
+        "Scaling learning rate from {} to {} due to bs factor {}".format(
+            config["lr"], lr, bs_factor
+        ),
+        flush=True,
+        )
+    
+    optimizer = optim.Adam(
+        [
+            {
+                "params": model.base_network.parameters(),
+                "lr": lr * config["lrAEmult"],
+            },
+            {"params": model.inner_model.parameters()},
+        ],  # model.parameters()
+        lr=lr,
+        betas=config["betas"],
+        eps=config["eps"],
+        weight_decay=config["weight_decay"],
+    )
+    if ("lr_annealingRate" not in config) or config[
+        "lr_annealingRate"
+    ] is None:
+        scheduler = None
+    else:
+        scheduler = torch.optim.lr_scheduler.StepLR(
+            optimizer, step_size=500, gamma=config["lr_annealingRate"]
+        )
+
+    return optimizer, scheduler, model
```

Comment:
This is going beyond what should be in a `model_factory` module, because they are parameters to the training, though it is OK for the `load_objects` functions. Rename the file.

Optimizer and LR scheduler should also be configurable in module_config, but with defaults available.